text/0000-item-level-scopes.md

```diff
@@ -0,0 +1,181 @@
+- Feature Name: Item-Level Scopes
+- Start Date: 2018-03-29
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Allow Items to be grouped into "Item-level scopes".
+
+# Motivation
+[motivation]: #motivation
+
+Sometimes one wishes to import a handful of functions/data structures from
+other modules, but only under some compile-time condition. For example, we
+might like to conditionally import some stuff only if we are in debug mode like
+this:
+
+```rust
+#[cfg(debug_assertions)]
+use a;
+#[cfg(debug_assertions)]
+use b;
+#[cfg(debug_assertions)]
+use c;
+#[cfg(debug_assertions)]
+use d
+```
+
+The above example is quite verbose because the `#[cfg...]` line must be
+repeated for each `use` line. This RFC proposes that the following should be
+possible:
+
+```rust
+#[cfg(debug_assertions)] {
+    use a;
+    use b;
+    use c;
+    use d;
+}
+```
+
+The latter is more concise and thus easier to read and write.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+An Item-level scope is allowed wherever an Item may appear. Item-level scopes
```

Comment:
I believe as currently framed, this RFC constitutes a breaking change.
Consider for example the following snippet which is valid today:

```rust
fn main () {
    #[cfg(debug_assertions)] {
        struct X;
    }
    #[cfg(debug_assertions)] {
        struct X;
    }
}
```

Here, the `struct X`s are items, but my understanding of this RFC is that `struct X` would both be exported to the parent scope (`fn main() { .. }`) and thus cause a conflict for existing code which the compiler doesn't know how to resolve.

The `mod { .. }` idea does not have this problem tho.

Reply:
That's true; However, it may cause confusion between the semantics of top level items and items inside functions. Meanwhile, `mod { .. }` as anonymous modules do not cause such confusion.